File: lib/services/StorageService.dart
// lib/services/StorageService.dart
import 'dart:io';
import 'package:firebase_storage/firebase_storage.dart';
import 'package:flutter/foundation.dart';
import 'package:uuid/uuid.dart';

import 'AuthService.dart';

class StorageService {
  final _storage = FirebaseStorage.instance;
  final _auth = AuthService();
  final _uuid = const Uuid();

  Future<String?> uploadPostImage(File file) async {
    try {
      final ownerId = _auth.currentUser?.uid;
      if (ownerId == null) return null;

      final ext = _ext(file.path);
      final path = 'posts/$ownerId/${DateTime.now().millisecondsSinceEpoch}_${_uuid.v4()}.$ext';
      final ref = _storage.ref().child(path);

      final meta = SettableMetadata(contentType: _contentType(ext));
      await ref.putFile(file, meta);
      return await ref.getDownloadURL();
    } catch (e) {
      debugPrint('uploadPostImage error: $e');
      return null;
    }
  }

  Future<List<String>> uploadPostImages(List<File> files, {String? uid}) async {
    final urls = <String>[];
    for (final f in files) {
      final u = await uploadPostImage(f);
      if (u != null) urls.add(u);
    }
    return urls;
    // 병렬 업로드 원하면:
    // final tasks = files.map((f) => uploadPostImage(f, uid: uid)).toList();
    // return (await Future.wait(tasks)).whereType<String>().toList();
  }

  Future<void> deleteByUrl(String url) async {
    try {
      await _storage.refFromURL(url).delete();
    } catch (e) {
      debugPrint('deleteByUrl error: $e');
    }
  }

  String _ext(String path) {
    final i = path.lastIndexOf('.');
    return (i >= 0 && i < path.length - 1) ? path.substring(i + 1).toLowerCase() : 'jpg';
  }

  String _contentType(String ext) {
    switch (ext) {
      case 'jpg':
      case 'jpeg':
        return 'image/jpeg';
      case 'png':
        return 'image/png';
      case 'webp':
        return 'image/webp';
      case 'gif':
        return 'image/gif';
      default:
        return 'application/octet-stream';
    }
  }
}

File: ting-rn/src/services/storageService.ts
import storage from '@react-native-firebase/storage';
import auth from '@react-native-firebase/auth';

function ext(path: string): string {
  const i = path.lastIndexOf('.');
  return i >= 0 && i < path.length - 1 ? path.substring(i + 1).toLowerCase() : 'jpg';
}

function contentType(e: string): string {
  switch (e) {
    case 'jpg':
    case 'jpeg':
      return 'image/jpeg';
    case 'png':
      return 'image/png';
    case 'webp':
      return 'image/webp';
    case 'gif':
      return 'image/gif';
    default:
      return 'application/octet-stream';
  }
}

export const storageService = {
  async uploadPostImage(fileUri: string): Promise<string | null> {
    const uid = auth().currentUser?.uid;
    if (!uid) return null;

    try {
      const fileExt = ext(fileUri);
      const path = `posts/${uid}/${Date.now()}_${Math.random().toString(36).slice(2)}.${fileExt}`;
      const ref = storage().ref().child(path);

      await ref.putFile(fileUri, { contentType: contentType(fileExt) });
      return ref.getDownloadURL();
    } catch {
      return null;
    }
  },

  async uploadPostImages(fileUris: string[]): Promise<string[]> {
    const results = await Promise.all(
      fileUris.map((uri) => this.uploadPostImage(uri)),
    );
    return results.filter((u): u is string => u !== null);
  },

  async deleteByUrl(url: string): Promise<void> {
    try {
      await storage().refFromURL(url).delete();
    } catch {
      // ignore
    }
  },
};
